### ISSUE TITLE
Why retiarii darts is much slower than legacy darts?

### ISSUE THREAD

**Issue by toufunao** [microsoft/nni#5103]
**Describe the issue**:
I used DDP to modify darts.py under /nni/retiarii/oneshot/pytorch/ and started the CNN training with search.py under nni/examples/nas/oneshot/darts. It took around **13min** to finish one epoch with 2 gpus. 

However, when I used the same CNN model and trained it with Retiarii, it took around **2h** to finish. It was much slower than expected. 

I noticed that Darts under Retiarii was implemented by Pytorch Lighting. Is it caused by Pytorch Lightning? 

Is it normal or is something wrong? 

Hope to get your help.

Attached are my codes. 

Retiarii:

`model = CNN(32, 3, args.channels, 10, args.layers)
    evaluator = pl.Classification(
   
      train_dataloaders=pl.DataLoader(train_set, batch_size=45,pin_memory=True,num_workers=4),
      val_dataloaders=pl.DataLoader(validate_set, batch_size=45,pin_memory=True,num_workers=4),
      
      max_epochs=1,
      accelerator="gpu", 
      devices=2,

      strategy='ddp',
      log_every_n_steps=10
    )
    exploration_strategy = strategy.DARTS()
    
    exp = RetiariiExperiment(model, evaluator, [], exploration_strategy)
    exp_config = RetiariiExeConfig('local')
    exp_config.experiment_name = 'cifa10_search'
    exp_config.max_trial_number = 1   
    exp_config.trial_concurrency = 1 
    exp_config.trial_gpu_number = 2
    exp_config.training_service.use_active_gpu = True

    exp_config.execution_engine = 'oneshot'
    
    exp.run(exp_config, 8081)`

Hint: 
In Retiarii, for each step, the architecture step used around 12s and the model step used around 12s as well.
In legacy, 25s is enough to finish 10 steps.

**Comment by ultmaster** [comment]
There is a known performance issue as of v2.8, which has already been addressed as of v2.9 (#5049).

Would you mind trying the test package of v2.9 and see whether it solves your issue?

```
pip install --extra-index-url https://test.pypi.org/simple/ nni==2.9a2
```

**Comment by toufunao** [comment]
Thanks. I tested again with the latest version and it was compatible, the time cost is merely the same.

**Comment by ultmaster** [comment]
I tested your code on 2 GTX 1080 GPUs. The estimated time of one epoch is around 13 minutes.

![image](https://user-images.githubusercontent.com/8463288/188059494-0a9b37d3-eb7f-4dc3-9713-2357f16857bc.png)

Here is my code for your reference.

```python
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

import json
import logging
import time
from argparse import ArgumentParser

import torch
import torch.nn as nn

import datasets
from model import CNN
from utils import accuracy


logger = logging.getLogger('nni')

if __name__ == "__main__":
    parser = ArgumentParser("darts")
    parser.add_argument("--layers", default=8, type=int)
    parser.add_argument("--batch-size", default=64, type=int)
    parser.add_argument("--log-frequency", default=10, type=int)
    parser.add_argument("--epochs", default=50, type=int)
    parser.add_argument("--channels", default=16, type=int)
    parser.add_argument("--unrolled", default=False, action="store_true")
    parser.add_argument("--visualization", default=False, action="store_true")
    parser.add_argument("--v1", default=False, action="store_true")
    args = parser.parse_args()

    dataset_train, dataset_valid = datasets.get_dataset("cifar10")

    model = CNN(32, 3, args.channels, 10, args.layers)
    criterion = nn.CrossEntropyLoss()

    optim = torch.optim.SGD(model.parameters(), 0.025, momentum=0.9, weight_decay=3.0E-4)
    lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, args.epochs, eta_min=0.001)

    from nni.retiarii.evaluator.pytorch import Classification, DataLoader

    model = CNN(32, 3, args.channels, 10, args.layers)
    evaluator = Classification(
        train_dataloaders=DataLoader(dataset_train, batch_size=45,pin_memory=True,num_workers=4),
        val_dataloaders=DataLoader(dataset_valid, batch_size=45,pin_memory=True,num_workers=4),
        max_epochs=1,
        accelerator="gpu", 
        devices=2,
        strategy='ddp',
        log_every_n_steps=10
    )

    from nni.retiarii.strategy import DARTS
    from nni.retiarii.experiment.pytorch import RetiariiExperiment, RetiariiExeConfig
    exploration_strategy = DARTS()

    exp = RetiariiExperiment(model, evaluator, [], exploration_strategy)
    exp_config = RetiariiExeConfig('local')
    exp_config.experiment_name = 'cifa10_search'
    exp_config.max_trial_number = 1   
    exp_config.trial_concurrency = 1 
    exp_config.trial_gpu_number = 2
    exp_config.training_service.use_active_gpu = True

    exp_config.execution_engine = 'oneshot'

    exp.run(exp_config, 8081)
```

**Comment by toufunao** [comment]
> I tested your code on 2 GTX 1080 GPUs. The estimated time of one epoch is around 13 minutes.
> 
> ![image](https://user-images.githubusercontent.com/8463288/188059494-0a9b37d3-eb7f-4dc3-9713-2357f16857bc.png)
> 
> Here is my code for your reference.
> 
> ```python
> # Copyright (c) Microsoft Corporation.
> # Licensed under the MIT license.
> 
> import json
> import logging
> import time
> from argparse import ArgumentParser
> 
> import torch
> import torch.nn as nn
> 
> import datasets
> from model import CNN
> from utils import accuracy
> 
> 
> logger = logging.getLogger('nni')
> 
> if __name__ == "__main__":
>     parser = ArgumentParser("darts")
>     parser.add_argument("--layers", default=8, type=int)
>     parser.add_argument("--batch-size", default=64, type=int)
>     parser.add_argument("--log-frequency", default=10, type=int)
>     parser.add_argument("--epochs", default=50, type=int)
>     parser.add_argument("--channels", default=16, type=int)
>     parser.add_argument("--unrolled", default=False, action="store_true")
>     parser.add_argument("--visualization", default=False, action="store_true")
>     parser.add_argument("--v1", default=False, action="store_true")
>     args = parser.parse_args()
> 
>     dataset_train, dataset_valid = datasets.get_dataset("cifar10")
> 
>     model = CNN(32, 3, args.channels, 10, args.layers)
>     criterion = nn.CrossEntropyLoss()
> 
>     optim = torch.optim.SGD(model.parameters(), 0.025, momentum=0.9, weight_decay=3.0E-4)
>     lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, args.epochs, eta_min=0.001)
> 
>     from nni.retiarii.evaluator.pytorch import Classification, DataLoader
> 
>     model = CNN(32, 3, args.channels, 10, args.layers)
>     evaluator = Classification(
>         train_dataloaders=DataLoader(dataset_train, batch_size=45,pin_memory=True,num_workers=4),
>         val_dataloaders=DataLoader(dataset_valid, batch_size=45,pin_memory=True,num_workers=4),
>         max_epochs=1,
>         accelerator="gpu", 
>         devices=2,
>         strategy='ddp',
>         log_every_n_steps=10
>     )
> 
>     from nni.retiarii.strategy import DARTS
>     from nni.retiarii.experiment.pytorch import RetiariiExperiment, RetiariiExeConfig
>     exploration_strategy = DARTS()
> 
>     exp = RetiariiExperiment(model, evaluator, [], exploration_strategy)
>     exp_config = RetiariiExeConfig('local')
>     exp_config.experiment_name = 'cifa10_search'
>     exp_config.max_trial_number = 1   
>     exp_config.trial_concurrency = 1 
>     exp_config.trial_gpu_number = 2
>     exp_config.training_service.use_active_gpu = True
> 
>     exp_config.execution_engine = 'oneshot'
> 
>     exp.run(exp_config, 8081)
> ```

Thanks, my testing result is the same as yours.

**Comment by ultmaster** [comment]
I'm a little curious what was the problem in your previous version.

**Comment by toufunao** [comment]
> 
In v2.8, training will take about 2 hours for just one epoch.

**Comment by ultmaster** [comment]
Closing as the problem seems to have been resolved in v2.9.

Please reopen if there are further inquiries.
